service/worker/scanner/scanner.go

```diff
@@ -161,23 +161,33 @@ func (s *Scanner) startShardScanner(
 	ctx context.Context,
 	config *shardscanner.ScannerConfig,
 ) (context.Context, []string) {
-	scannerContextKey := shardscanner.ScannerContextKey(config.ScannerWFTypeName)
-	fixerContextKey := shardscanner.ScannerContextKey(config.FixerWFTypeName)
 	workerTaskListNames := []string{}
-	backgroundActivityContext := context.WithValue(ctx, scannerContextKey, s.context)
-
 	if config.DynamicParams.ScannerEnabled() {
-		backgroundActivityContext = context.WithValue(
-			backgroundActivityContext,
-			scannerContextKey,
+		ctx = context.WithValue(
+			ctx,
+			shardscanner.ScannerContextKey(config.ScannerWFTypeName),
```

Comment:
I am a bit worried about the way of managing contextKeys. Two potential issues here --
1. They could conflict each other because the next person may not be able to find all the existing keys already used
2. The bug will come back again on another refactoring. Before we can implement any test(it's quite tricky and I don't even have a good idea to test yet), may be just write a code in a more organized way:

So How about having a separate file/module to define all the context keys in a `map[string]interface{}`, and then return an error when starting scanner if any key is not set in the backgroundContext?

Reply:
yea, personally I dislike exposing keys _ever_.  e.g. rather than `context.WithValue(ctx, getKeyFunc(..), value)`, do `setValue(ctx, key_args, value)` + a typed get-value func.  makes it effectively impossible to misuse, trivial to find all calls, you can add validation -> return err, etc.

Reply:
yeah dislike those global keys too. That's the main reasons that we ran into the bug twice, so we should improve this piece of code. 
 
Not 100% following what you are suggesting but seems interesting. Are you saying having a method `setValue` of Scanner to manage that?

Reply:
basically:
- all keys should be private + uniquely typed (+ personally: strings, so they're readable in logs/debuggers).  that way the value does not matter.
- in that same file: make as-strongly-typed-as-possible setter and getter funcs, which read/write/cast the value with the keys
- literally nothing except those funcs ever uses the keys, ideally guaranteed by the compiler because they're private
- since everything's trivial and in one file, you can verify it at a glance and move on with your life

that way it's all typed, you can add validation in insertion and/or retrieval, and there's zero chance of collisions or wrong type casts.

Reply:
Just to add more context here, we are dealing here with two separate context types:

- shardscanner context which is using typed string for context key. While adding new shardscanner, you don't even have to think about context as it uses workflow name as context key,  which also helps to avoid collisions as you are not allowed to have different workflows under the same name.

- TaskListScanner/HistoryScanner context which shares the same context key provided by a constant which usage was mistakenly moved to another scope]